cf-custom-resources/lib/wkld-cert-validator.js

```diff
@@ -0,0 +1,881 @@
+          break;
+        }
+        let unusedOptions = await unusedValidationOptions(
+          physicalResourceID,
+          loadBalancerDNS
+        );
+        await devalidate(unusedOptions);
+        await deleteCertificate(physicalResourceID);
+        break;
+      default:
+        throw new Error(`Unsupported request type ${event.RequestType}`);
+    }
+  };
+
+  try {
+    await Promise.race([exports.deadlineExpired(), handler()]);
+    await report(event, context, "SUCCESS", physicalResourceID);
+  } catch (err) {
+    console.log(`Caught error for service ${serviceName}: ${err.message}`);
+    await report(
+      event,
+      context,
+      "FAILED",
+      physicalResourceID,
+      null,
+      err.message
+    );
+  }
+};
+
+/**
+ * Delete the certificate.
+ *
+ * @param certARN The ARN of the certificate to delete.
+ * @returns {Promise<void>}
+ */
+async function deleteCertificate(certARN) {
+  // NOTE: wait for certificate to be not in-used.
+  let attempt;
+  for (attempt = 0; attempt < ATTEMPTS_CERTIFICATE_NOT_IN_USE; attempt++) {
+    let certificate;
+    try {
+      ({ Certificate: certificate } = await clients
+        .acm()
+        .describeCertificate({
+          CertificateArn: certARN,
+        })
+        .promise());
+    } catch (err) {
+      if (err.name === "ResourceNotFoundException") {
+        return;
+      }
+      throw err;
+    }
+
+    if (!certificate.InUseBy || certificate.InUseBy.length <= 0) {
+      break;
+    }
+    await sleep(30000);
+  }
+
+  if (attempt >= ATTEMPTS_CERTIFICATE_NOT_IN_USE) {
+    throw new Error(
+      `Certificate still in use after checking for ${ATTEMPTS_CERTIFICATE_NOT_IN_USE} attempts.`
+    );
+  }
+
+  await clients
+    .acm()
+    .deleteCertificate({ CertificateArn: certARN })
+    .promise()
+    .catch((err) => {
+      if (err.name !== "ResourceNotFoundException") {
+        throw err;
+      }
+    });
+}
+
+/**
+ * Validate that the aliases are not in use.
+ *
+ * @param {Set<String>} aliases for the service.
+ * @param {String} loadBalancerDNS the DNS of the service's load balancer.
+ * @throws error if at least one of the aliases is not valid.
+ */
+async function validateAliases(aliases, loadBalancerDNS) {
+  let promises = [];
+
+  for (let alias of aliases) {
+    let { hostedZoneID, route53Client } = await domainResources(alias);
+    const promise = route53Client
+      .listResourceRecordSets({
+        HostedZoneId: hostedZoneID,
+        MaxItems: "1",
+        StartRecordName: alias,
+        StartRecordType: "A",
+      })
+      .promise()
+      .then(({ ResourceRecordSets: recordSet }) => {
+        if (!targetRecordExists(alias, recordSet)) {
+          return;
+        }
+        if (recordSet[0].Type !== "A") {
+          return;
+        }
+        let aliasTarget = recordSet[0].AliasTarget;
+        if (
+          aliasTarget &&
+          loadBalancerDNS &&
```

Comment:
Do we not want to make sure that the alias is pointing to the correct CF DNS as well?

Reply:
So static site alias is different because we won't be switching the alias destination between CF/ALB. Also from implementation perspective unless we write a Custom Resource to update the CF, we won't be able to do this as the certificate is required by CF creation (we wouldn't know the CF DNS name in this lambda)

Reply:
What would happen in this case ⬇️ 
```
// A LBWS
http:
  alias: svc.example.com

// A static site
http:
  alias: svc.example.com # Mistakenly wrote the same alias here
```

Originally, the purpose of this check was to make sure that the alias record that we are trying to update/delete is indeed owned by this particular service. We didn't want to update the alias record if it was pointing to something unexpected. I feel like the same concern applies to static site 🤔

Reply:
I think a key difference for static site that makes this not a concern is the fact that this custom resource won't update A record as it does in LBWS. I'll do more manual test to make sure this is not an issue.

Reply:
Discussed offline will add more comment justifying why we are doing this.